msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java

```diff
@@ -2224,6 +2236,220 @@ public String processActionToggleDisplayExtendedDescription()
+          ((DiscussionForum) forumItem).setCalendarBeginId(null);
+          ((DiscussionForum) forumItem).setCalendarEndId(null);
+        }
+      }
+    } catch (IdUnusedException e) { //for getCalendar, if the Calendar doesn't exist.
+      e.printStackTrace();
+    } catch (PermissionException e) { //for addEvent, if the user can't edit a calendar.
+      e.printStackTrace();
+    } catch (InUseException e) {  //for getEditEvent, if it's already being edited.
+      e.printStackTrace();
+    }
+  }
+
+  public Boolean getDoesSiteHaveCalendar(){  //if any Calendar data exists for this site, this method returns True. It's not necessarily based on the literal presence of the Calendar tool in the site, but on Calendar-related data.
+    try { //just look for a usable calendar with this site's ID.
+      this.calendarService.getCalendar("/calendar/calendar/" + getContextSiteId().replace("/site/", "") + "/main");
+    } catch (IdUnusedException e) { //this would be the normal Catch case...when the calendar we tried to get just doesn't exist.
+      log.debug(" No Calendar data exist for this site.");
+      return false;
+    } catch (PermissionException e) { //this just means the user can't edit. Probably rare.
+      log.debug(" User does not have permission to edit this calendar.");
+      return false;
+    }
+    return true;  // if no exceptions get thrown, return True...either usable Calendar data exists, or the Calendar tool is in the site.
+  }
+
+  private Collection<Group> getAllowedGroups(Set membershipItems) {  //get the groups allowed to see a Forum item's related Calendar events.
+    if (membershipItems == null) {
+      return new ArrayList<>();
+    }
+    Iterator traverseMembershipItems = membershipItems.iterator();
+    Collection<Group> output = new ArrayList<Group>();
+    ArrayList<SiteGroupBean> availableGroups = (ArrayList<SiteGroupBean>) getSiteGroups();
+    ArrayList<String> availableGroupNames = new ArrayList<String>();
+    for (int count = 0; count < availableGroups.size(); count++){ //build an array of just the group names.
+      availableGroupNames.add(availableGroups.get(count).getGroup().getTitle());
+    }
+    while (traverseMembershipItems.hasNext()) {
+      DBMembershipItemImpl now = (DBMembershipItemImpl) traverseMembershipItems.next();
+      if (availableGroupNames.contains(now.getName()) && !now.getPermissionLevelName().equals("None")) { //we want the groups that are available to the site, and have some kind of permission to see the forum item.
+        for (int count = 0; count < availableGroups.size(); count++) {
+          if (now.getName().equals(availableGroups.get(count).getGroup().getTitle())){
+            output.add(availableGroups.get(count).getGroup());
+          }
+        }
+      }
+    }
+    return output;
+  }
+
+  private Boolean studentsAllowed (Set membershipItems){  //find out if users in the Student role are allowed to use a Forum Item in any way.
```

Comment:
You should type all this and try and use streams and lambdas to make it more concise.